From d738eb47a3e781477101d706f01dd61148a9086f Mon Sep 17 00:00:00 2001
From: Alan Coopersmith <alan.coopersmith@oracle.com>
Date: Fri, 1 Mar 2013 20:54:24 -0800
Subject: [PATCH:libX11 02/38] Add _XEatDataWords to discard a given number of
 32-bit words of reply data

Matches the units of the length field in X protocol replies, and provides
a single implementation of overflow checking to avoid having to replicate
those checks in every caller.

Signed-off-by: Alan Coopersmith <alan.coopersmith@oracle.com>
Reviewed-by: Matthieu Herrb <matthieu.herrb@laas.fr>
---
 include/X11/Xlibint.h |    4 ++++
 src/xcb_io.c          |   17 +++++++++++++++++
 2 files changed, 21 insertions(+)

diff --git a/include/X11/Xlibint.h b/include/X11/Xlibint.h
index 06395b3..d63a534 100644
--- a/include/X11/Xlibint.h
+++ b/include/X11/Xlibint.h
@@ -855,6 +855,10 @@ extern void _XEatData(
     Display*		/* dpy */,
     unsigned long	/* n */
 );
+extern void _XEatDataWords(
+    Display*		/* dpy */,
+    unsigned long	/* n */
+);
 extern char *_XAllocScratch(
     Display*		/* dpy */,
     unsigned long	/* nbytes */
diff --git a/src/xcb_io.c b/src/xcb_io.c
index 300ef57..727c6c7 100644
--- a/src/xcb_io.c
+++ b/src/xcb_io.c
@@ -19,6 +19,7 @@
 #include <stdint.h>
 #include <stdlib.h>
 #include <string.h>
+#include <limits.h>
 #ifdef HAVE_SYS_SELECT_H
 #include <sys/select.h>
 #endif
@@ -757,3 +758,19 @@ void _XEatData(Display *dpy, unsigned long n)
 	dpy->xcb->reply_consumed += n;
 	_XFreeReplyData(dpy, False);
 }
+
+/*
+ * Read and discard "n" 32-bit words of data
+ * Matches the units of the length field in X protocol replies, and provides
+ * a single implementation of overflow checking to avoid having to replicate
+ * those checks in every caller.
+ */
+void _XEatDataWords(Display *dpy, unsigned long n)
+{
+	if (n < ((INT_MAX - dpy->xcb->reply_consumed) >> 2))
+		dpy->xcb->reply_consumed += (n << 2);
+	else
+		/* Overflow would happen, so just eat the rest of the reply */
+		dpy->xcb->reply_consumed = dpy->xcb->reply_length;
+	_XFreeReplyData(dpy, False);
+}
-- 
1.7.9.2

From 56623594bb0f132aa2792f76378a573c7efe31e2 Mon Sep 17 00:00:00 2001
From: Alan Coopersmith <alan.coopersmith@oracle.com>
Date: Fri, 19 Apr 2013 14:30:40 -0700
Subject: [PATCH:libX11 38/38] Give GNU & Solaris Studio compilers hints about
 XEatData branches

Try to offset the cost of all the recent checks we've added by giving
the compiler a hint that the branches that involve us eating data
are less likely to be used than the ones that process it.

Signed-off-by: Alan Coopersmith <alan.coopersmith@oracle.com>
---
 include/X11/Xlibint.h |   16 ++++++++++++++--
 1 file changed, 14 insertions(+), 2 deletions(-)

diff --git a/include/X11/Xlibint.h b/include/X11/Xlibint.h
index d63a534..acbad6b 100644
--- a/include/X11/Xlibint.h
+++ b/include/X11/Xlibint.h
@@ -832,6 +832,15 @@ typedef struct _XExten {		/* private to extension mechanism */
 	struct _XExten *next_flush;	/* next in list of those with flushes */
 } _XExtension;
 
+/* Temporary definition until we can depend on an xproto release with it */
+#ifdef _X_COLD
+# define _XLIB_COLD _X_COLD
+#elif defined(__GNUC__) && ((__GNUC__ * 100 + __GNUC_MINOR__) >= 403) /* 4.3+ */
+# define _XLIB_COLD __attribute__((__cold__))
+#else
+# define _XLIB_COLD /* nothing */
+#endif
+
 /* extension hooks */
 
 #ifdef DataRoutineIsProcedure
@@ -854,11 +863,14 @@ extern int (*_XErrorFunction)(
 extern void _XEatData(
     Display*		/* dpy */,
     unsigned long	/* n */
-);
+) _XLIB_COLD;
 extern void _XEatDataWords(
     Display*		/* dpy */,
     unsigned long	/* n */
-);
+) _XLIB_COLD;
+#if defined(__SUNPRO_C) /* Studio compiler alternative to "cold" attribute */
+# pragma rarely_called(_XEatData, _XEatDataWords)
+#endif
 extern char *_XAllocScratch(
     Display*		/* dpy */,
     unsigned long	/* nbytes */
-- 
1.7.9.2

From d0944b52f8debb74eb48359a8dc20706b12da834 Mon Sep 17 00:00:00 2001
From: Alan Coopersmith <alan.coopersmith@oracle.com>
Date: Sat, 2 Mar 2013 16:56:16 -0800
Subject: [PATCH:libX11 34/38] Convert more _XEatData callers to
 _XEatDataWords

Signed-off-by: Alan Coopersmith <alan.coopersmith@oracle.com>
Reviewed-by: Matthieu Herrb <matthieu.herrb@laas.fr>
---
 src/GetAtomNm.c |    4 ++--
 src/LiICmaps.c  |    8 ++++----
 src/LiProps.c   |    8 ++++----
 src/OpenDis.c   |    2 +-
 src/QuColors.c  |   10 +++++-----
 src/QuTree.c    |    8 ++++----
 6 files changed, 20 insertions(+), 20 deletions(-)

diff --git a/src/GetAtomNm.c b/src/GetAtomNm.c
index 9823c69..996f7eb 100644
--- a/src/GetAtomNm.c
+++ b/src/GetAtomNm.c
@@ -78,7 +78,7 @@ char *XGetAtomName(
 	name[rep.nameLength] = '\0';
 	_XUpdateAtomCache(dpy, name, atom, 0, -1, 0);
     } else {
-	_XEatData(dpy, (unsigned long) (rep.nameLength + 3) & ~3);
+	_XEatDataWords(dpy, rep.length);
 	name = (char *) NULL;
     }
     UnlockDisplay(dpy);
@@ -176,7 +176,7 @@ XGetAtomNames (
 		_XUpdateAtomCache(dpy, names_return[missed], atoms[missed],
 				  0, -1, 0);
 	    } else {
-		_XEatData(dpy, (unsigned long) (rep.nameLength + 3) & ~3);
+		_XEatDataWords(dpy, rep.length);
 		async_state.status = 0;
 	    }
 	}
diff --git a/src/LiICmaps.c b/src/LiICmaps.c
index e981619..45a2f2f 100644
--- a/src/LiICmaps.c
+++ b/src/LiICmaps.c
@@ -34,7 +34,7 @@ Colormap *XListInstalledColormaps(
     Window win,
     int *n)  /* RETURN */
 {
-    long nbytes;
+    unsigned long nbytes;
     Colormap *cmaps;
     xListInstalledColormapsReply rep;
     register xResourceReq *req;
@@ -51,14 +51,14 @@ Colormap *XListInstalledColormaps(
 
     if (rep.nColormaps) {
 	nbytes = rep.nColormaps * sizeof(Colormap);
-	cmaps = (Colormap *) Xmalloc((unsigned) nbytes);
-	nbytes = rep.nColormaps << 2;
+	cmaps = Xmalloc(nbytes);
 	if (! cmaps) {
-	    _XEatData(dpy, (unsigned long) nbytes);
+	    _XEatDataWords(dpy, rep.length);
 	    UnlockDisplay(dpy);
 	    SyncHandle();
 	    return((Colormap *) NULL);
 	}
+	nbytes = rep.nColormaps << 2;
 	_XRead32 (dpy, (long *) cmaps, nbytes);
     }
     else cmaps = (Colormap *) NULL;
diff --git a/src/LiProps.c b/src/LiProps.c
index 72560ab..d9c7465 100644
--- a/src/LiProps.c
+++ b/src/LiProps.c
@@ -34,7 +34,7 @@ Atom *XListProperties(
     Window window,
     int *n_props)  /* RETURN */
 {
-    long nbytes;
+    unsigned long nbytes;
     xListPropertiesReply rep;
     Atom *properties;
     register xResourceReq *req;
@@ -50,14 +50,14 @@ Atom *XListProperties(
 
     if (rep.nProperties) {
 	nbytes = rep.nProperties * sizeof(Atom);
-	properties = (Atom *) Xmalloc ((unsigned) nbytes);
-	nbytes = rep.nProperties << 2;
+	properties = Xmalloc (nbytes);
 	if (! properties) {
-	    _XEatData(dpy, (unsigned long) nbytes);
+	    _XEatDataWords(dpy, rep.length);
 	    UnlockDisplay(dpy);
 	    SyncHandle();
 	    return (Atom *) NULL;
 	}
+	nbytes = rep.nProperties << 2;
 	_XRead32 (dpy, (long *) properties, nbytes);
     }
     else properties = (Atom *) NULL;
diff --git a/src/OpenDis.c b/src/OpenDis.c
index f6d8c70..0bf1b91 100644
--- a/src/OpenDis.c
+++ b/src/OpenDis.c
@@ -552,7 +552,7 @@ XOpenDisplay (
 		    dpy->xdefaults[reply.nItems] = '\0';
 		}
 		else if (reply.propertyType != None)
-		    _XEatData(dpy, reply.nItems * (reply.format >> 3));
+		    _XEatDataWords(dpy, reply.length);
 	    }
 	}
 	UnlockDisplay(dpy);
diff --git a/src/QuColors.c b/src/QuColors.c
index 237b8bf..13a63eb 100644
--- a/src/QuColors.c
+++ b/src/QuColors.c
@@ -37,9 +37,7 @@ _XQueryColors(
     int ncolors)
 {
     register int i;
-    xrgb *color;
     xQueryColorsReply rep;
-    long nbytes;
     register xQueryColorsReq *req;
 
     GetReq(QueryColors, req);
@@ -52,8 +50,9 @@ _XQueryColors(
        /* XXX this isn't very efficient */
 
     if (_XReply(dpy, (xReply *) &rep, 0, xFalse) != 0) {
-	if ((color = (xrgb *)
-	    Xmalloc((unsigned) (nbytes = (long) ncolors * SIZEOF(xrgb))))) {
+	unsigned long nbytes = (long) ncolors * SIZEOF(xrgb);
+	xrgb *color = Xmalloc(nbytes);
+	if (color != NULL) {
 
 	    _XRead(dpy, (char *) color, nbytes);
 
@@ -67,7 +66,8 @@ _XQueryColors(
 	    }
 	    Xfree((char *)color);
 	}
-	else _XEatData(dpy, (unsigned long) nbytes);
+	else
+	    _XEatDataWords(dpy, rep.length);
     }
 }
 
diff --git a/src/QuTree.c b/src/QuTree.c
index 3cea282..8da2ae2 100644
--- a/src/QuTree.c
+++ b/src/QuTree.c
@@ -37,7 +37,7 @@ Status XQueryTree (
     Window **children,	/* RETURN */
     unsigned int *nchildren)  /* RETURN */
 {
-    long nbytes;
+    unsigned long nbytes;
     xQueryTreeReply rep;
     register xResourceReq *req;
 
@@ -52,14 +52,14 @@ Status XQueryTree (
     *children = (Window *) NULL;
     if (rep.nChildren != 0) {
 	nbytes = rep.nChildren * sizeof(Window);
-	*children = (Window *) Xmalloc((unsigned) nbytes);
-	nbytes = rep.nChildren << 2;
+	*children = Xmalloc(nbytes);
 	if (! *children) {
-	    _XEatData(dpy, (unsigned long) nbytes);
+	    _XEatDataWords(dpy, rep.length);
 	    UnlockDisplay(dpy);
 	    SyncHandle();
 	    return (0);
 	}
+	nbytes = rep.nChildren << 2;
 	_XRead32 (dpy, (long *) *children, nbytes);
     }
     *parent = rep.parent;
-- 
1.7.9.2

